emote/models/ensemble.py

```diff
@@ -0,0 +1,351 @@
+        logvar = self.max_logvar - F.softplus(self.max_logvar - logvar)
+        logvar = self.min_logvar + F.softplus(logvar - self.min_logvar)
+        return mean, logvar
+
+    def _forward_from_indices(
+        self, x: torch.Tensor, model_shuffle_indices: torch.Tensor
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        _, batch_size, _ = x.shape
+
+        num_models = (
+            len(self.elite_models) if self.elite_models is not None else len(self)
+        )
+        shuffled_x = x[:, model_shuffle_indices, ...].view(
+            num_models, batch_size // num_models, -1
+        )
+
+        mean, logvar = self._default_forward(shuffled_x, only_elite=True)
+        # note that mean and logvar are shuffled
+        mean = mean.view(batch_size, -1)
+        mean[model_shuffle_indices] = mean.clone()  # invert the shuffle
+
+        if logvar is not None:
+            logvar = logvar.view(batch_size, -1)
+            logvar[model_shuffle_indices] = logvar.clone()  # invert the shuffle
+
+        return mean, logvar
+
+    def _forward_ensemble(
+        self, x: torch.Tensor
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """
+        Valid propagation options are:
+            - "random_model": for each output in the batch a model will be chosen at random.
+              This corresponds to TS1 propagation in the PETS paper.
+            - "fixed_model": for output j-th in the batch, the model will be chosen according to
+              the model index in `propagation_indices[j]`. This can be used to implement TSinf
+              propagation, described in the PETS paper.
+            - "expectation": the output for each element in the batch will be the mean across
+              models.
+        """
+        if self.propagation_method is None:
+            mean, logvar = self._default_forward(x, only_elite=False)
+            if self.num_members == 1:
+                mean = mean[0]
+                logvar = logvar[0] if logvar is not None else None
+            return mean, logvar
+        assert x.ndim == 2
+        model_len = (
+            len(self.elite_models) if self.elite_models is not None else len(self)
+        )
+        if x.shape[0] % model_len != 0:
+            raise ValueError(
+                f"GaussianMLP ensemble requires batch size to be a multiple of the "
+                f"number of models. Current batch size is {x.shape[0]} for "
+                f"{model_len} models."
+            )
+        x = x.unsqueeze(0)
+        if self.propagation_method == "random_model":
+            # passing generator causes segmentation fault
+            # see https://github.com/pytorch/pytorch/issues/44714
+            model_indices = torch.randperm(x.shape[1], device=self.device)
+            return self._forward_from_indices(x, model_indices)
+        if self.propagation_method == "fixed_model":
+            if self.propagation_indices is None:
```

Comment:
Can be checked in init?

Reply:
We will only use expectation propagation, and all other methods are removed.
Fixed then!